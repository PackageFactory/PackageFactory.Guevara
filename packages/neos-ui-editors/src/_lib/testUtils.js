import {PureComponent, Children} from 'react';
import PropTypes from 'prop-types';
import {SynchronousRegistry, SynchronousMetaRegistry} from '@neos-project/neos-ui-extensibility/src/registry';

const globalRegistry = new SynchronousMetaRegistry();

// i18n Registry
class FakeI18NRegistry extends SynchronousRegistry {
    translate(key) {
        return key;
    }
}
globalRegistry.set('i18n', new FakeI18NRegistry());

// Data Source DataLoader
export const MockDataSourceDataLoader = {
    _currentPromise: null,
    _currentPromiseResolveFn: null,

    // API in tests
    reset() {
        this._currentPromise = null;
    },

    resolveCurrentPromise(result) {
        this._currentPromiseResolveFn(result);
        return this._currentPromise;
    },

    // API of DataLoader
    resolveValue() {
        if (!this._currentPromise) {
            this._currentPromise = new Promise(resolve => {
                this._currentPromiseResolveFn = resolve;
            });
        }
        return this._currentPromise;
    }
};
globalRegistry.set('dataLoaders', new SynchronousRegistry());
globalRegistry.get('dataLoaders').set('DataSources', MockDataSourceDataLoader);

// wrapper
export class WrapWithMockGlobalRegistry extends PureComponent {
    static propTypes = {
        children: PropTypes.element.isRequired
    };

    static childContextTypes = {
        globalRegistry: PropTypes.object.isRequired,
        configuration: PropTypes.object.isRequired
    };

    getChildContext() {
        const configuration = {};
        return {globalRegistry, configuration};
    }

    render() {
        return Children.only(this.props.children);
    }
}
