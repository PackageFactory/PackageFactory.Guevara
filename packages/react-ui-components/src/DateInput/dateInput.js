import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import enhanceWithClickOutside from 'react-click-outside';
import moment from 'moment';
import mergeClassNames from 'classnames';

export class DateInput extends PureComponent {
    static propTypes = {
        /**
         * The Date instance which represents the selected value.
         */
        value: PropTypes.instanceOf(Date),

        /**
         * An optional placeholder which will be rendered if no date was selected.
         */
        placeholder: PropTypes.string,

        /**
         * An optional id for the input.
         */
        id: PropTypes.string,

        /**
         * The label which will be displayed within the `Select Today` btn.
         */
        todayLabel: PropTypes.string,

        /**
         * The label which will be displayed within the `Select Today` btn.
         */
        applyLabel: PropTypes.string,

        /**
         * The moment format string to use to format the passed value.
         */
        labelFormat: PropTypes.string,

        /**
         * Highlight input
         */
        highlight: PropTypes.bool,

        /**
         * Display only date picker
         */
        dateOnly: PropTypes.bool,

        /**
         * Display only time picker
         */
        timeOnly: PropTypes.bool,

        /**
         * The changehandler to call when the date changes.
         */
        onChange: PropTypes.func.isRequired,
        theme: PropTypes.shape({/* eslint-disable quote-props */
            'wrapper': PropTypes.string,
            'calendarInputWrapper': PropTypes.string,
            'calendarIconBtn': PropTypes.string,
            'calendarFakeInputWrapper': PropTypes.string,
            'calendarFakeInputMirror': PropTypes.string,
            'calendarFakeInput': PropTypes.string,
            'closeCalendarIconBtn': PropTypes.string,
            'selectTodayBtn': PropTypes.string
        }).isRequired, /* eslint-enable quote-props */

        //
        // Static component dependencies which are injected from the outside (index.js)
        //
        ButtonComponent: PropTypes.any.isRequired,
        IconComponent: PropTypes.any.isRequired,
        DatePickerComponent: PropTypes.any.isRequired,
        CollapseComponent: PropTypes.any.isRequired
    };

    static defaultProps = {
        todayLabel: 'Today',
        applyLabel: 'Apply',
        labelFormat: 'DD-MM-YYYY hh:mm'
    };

    constructor(props) {
        super(props);

        this.state = {
            isOpen: false,
            transientDate: null
        };
        this.handleChange = this.handleChange.bind(this);
        this.handleApply = this.handleApply.bind(this);
        this.handleClearValueClick = this.handleClearValueClick.bind(this);
        this.handleInputClick = this.open.bind(this);
        this.handleCalendarIconClick = this.open.bind(this);
        this.handleClickOutside = this.close.bind(this);
        this.handleSelectTodayBtnClick = this.handleSelectTodayBtnClick.bind(this);
    }

    render() {
        const {
            ButtonComponent,
            IconComponent,
            DatePickerComponent,
            CollapseComponent,
            placeholder,
            theme,
            value,
            id,
            applyLabel,
            todayLabel,
            labelFormat,
            dateOnly,
            timeOnly,
            highlight
        } = this.props;
        const selectedDate = value ? moment(value).format(labelFormat) : '';

        const calendarInputWrapper = mergeClassNames({
            [theme.calendarInputWrapper]: true,
            [theme['calendarInputWrapper--highlight']]: highlight
        });

        return (
            <div className={theme.wrapper}>
                <div className={calendarInputWrapper}>
                    <button
                        onClick={this.handleCalendarIconClick}
                        className={theme.calendarIconBtn}
                        >
                        <IconComponent icon="calendar"/>
                    </button>
                    <div className={theme.calendarFakeInputWrapper}>
                        <div
                            role="presentation"
                            onClick={this.handleInputClick}
                            className={theme.calendarFakeInputMirror}
                            />
                        <input
                            id={id}
                            onFocus={this.handleInputClick}
                            type="datetime"
                            placeholder={placeholder}
                            className={theme.calendarFakeInput}
                            value={selectedDate}
                            readOnly
                            />
                    </div>
                    <button
                        onClick={this.handleClearValueClick}
                        className={theme.closeCalendarIconBtn}
                        >
                        <IconComponent icon="remove"/>
                    </button>
                </div>
                <CollapseComponent isOpened={this.state.isOpen}>
                    <button
                        className={theme.selectTodayBtn}
                        onClick={this.handleSelectTodayBtnClick}
                        >
                        {todayLabel}
                    </button>
                    <DatePickerComponent
                        open={true}
                        defaultValue={value}
                        dateFormat={!timeOnly}
                        timeFormat={!dateOnly}
                        onChange={this.handleChange}
                        />
                    <ButtonComponent
                        onClick={this.handleApply}
                        className={theme.applyBtn}
                        style="brand"
                        >
                        {applyLabel}
                    </ButtonComponent>
                </CollapseComponent>
            </div>
        );
    }

    handleChange(momentVal) {
        const date = momentVal.toDate();
        this.setState({
            transientDate: date
        });
    }

    handleApply() {
        this.setState({
            isOpen: false
        }, () => {
            this.props.onChange(this.state.transientDate);
        });
    }

    handleClearValueClick() {
        this.setState({
            isOpen: false
        }, () => {
            this.props.onChange(null);
        });
    }

    handleSelectTodayBtnClick() {
        this.setState({
            isOpen: false
        }, () => {
            this.props.onChange(moment().toDate());
        });
    }

    open() {
        this.setState({
            isOpen: true
        });
    }

    close() {
        this.setState({
            isOpen: false
        });
    }
}

//
// Add the click-outside functionality to the DateInput component.
//
const EnhancedDateInput = enhanceWithClickOutside(DateInput);

export default EnhancedDateInput;
