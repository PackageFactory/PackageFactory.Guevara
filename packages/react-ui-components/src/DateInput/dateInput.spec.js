import sinon from 'sinon';
import moment from 'moment';
import {createShallowRenderer, createStubComponent} from './../_lib/testUtils.js';
import {DateInput} from './dateInput.js';

const defaultProps = {
    onChange: () => null,
    IconComponent: createStubComponent(),
    DatePickerComponent: createStubComponent(),
    CollapseComponent: createStubComponent(),
    theme: {
        'dropDown__contents': 'baseDropDownContentsClassName',
        'dropDown__contents--isOpen': 'openDropDownContentsClassName'
    }
};
const shallow = createShallowRenderer(DateInput, defaultProps);

test('should initialize with a state of {isOpen = false}.', () => {
    const wrapper = shallow();

    expect(wrapper.state('isOpen')).toBe(false);
});
test('should render 3 buttons.', () => {
    const wrapper = shallow();
    const btns = wrapper.find('button');

    expect(btns.length).toBe(3);
});
test('should render an calendar icon within the first button.', () => {
    const wrapper = shallow();
    const btn = wrapper.find('button').at(0);

    expect(btn.find(defaultProps.IconComponent).length).toBe(1);
    expect(btn.find(defaultProps.IconComponent).prop('icon')).toBe('calendar');
});
test('should set the "isOpen" state to a truthy value when clicking on the first button.', () => {
    const wrapper = shallow();
    const btn = wrapper.find('button').at(0);

    btn.simulate('click');

    expect(wrapper.state('isOpen')).toBe(true);
});
test('should render a input[type="datetime"].', () => {
    const wrapper = shallow();
    const input = wrapper.find('input[type="datetime"]');

    input.simulate('focus');

    expect(input.length).toBe(1);
    expect(wrapper.state('isOpen')).toBe(true);
});
test('should render an remove icon within the second button.', () => {
    const wrapper = shallow();
    const btn = wrapper.find('button').at(1);

    expect(btn.find(defaultProps.IconComponent).length).toBe(1);
    expect(btn.find(defaultProps.IconComponent).prop('icon')).toBe('remove');
});
test('should set the "isOpen" state to a falsy value and call the "onChange" prop with `null` when clicking on the second button.', () => {
    const onChange = sinon.spy();
    const wrapper = shallow({onChange});
    const btn = wrapper.find('button').at(1);

    wrapper.setState({isOpen: true});
    btn.simulate('click');

    expect(wrapper.state('isOpen')).toBe(false);
    expect(onChange.callCount).toBe(1);
    expect(onChange.args[0][0]).toBe(null);
});
test('should render a "Collapse" Component which is opened depending on the "isOpen" state.', () => {
    const wrapper = shallow();
    const collapse = wrapper.find(defaultProps.CollapseComponent);

    expect(collapse.length).toBe(1);
    expect(collapse.prop('isOpened')).toBe(wrapper.state('isOpen'));
});
test('should render a "DatePicker" Component.', () => {
    const value = new Date();
    const wrapper = shallow({value});
    const picker = wrapper.find(defaultProps.DatePickerComponent);

    expect(picker.length).toBe(1);
    expect(picker.prop('open')).toBe(true);
    expect(picker.prop('value')).toBe(value);
});
test('should call the "onChange" prop when triggering the change event on the "DatePicker" Component.', () => {
    const onChange = sinon.spy();
    const value = new Date();
    const wrapper = shallow({value, onChange});
    const picker = wrapper.find(defaultProps.DatePickerComponent);
    const newVal = moment();

    picker.simulate('change', newVal);

    expect(onChange.callCount).toBe(1);
    expect(onChange.args[0][0].toTimeString()).toBe(newVal.toDate().toTimeString());
});
test('should set the "isOpen" state to a falsy value and call the "onChange" prop with todays date when clicking on the third button.', () => {
    const onChange = sinon.spy();
    const wrapper = shallow({onChange});
    const btn = wrapper.find('button').at(2);
    const date = new Date();

    wrapper.setState({isOpen: true});
    btn.simulate('click');

    expect(wrapper.state('isOpen')).toBe(false);
    expect(onChange.callCount).toBe(1);
    expect(onChange.args[0][0].toTimeString()).toBe(date.toTimeString());
});
