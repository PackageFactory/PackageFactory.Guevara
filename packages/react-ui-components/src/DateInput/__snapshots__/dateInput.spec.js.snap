// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateInput/> should render correctly. 1`] = `
<div>
  <div
    className=""
  >
    <button
      onClick={[Function]}
    >
      <Component
        icon="calendar"
      />
    </button>
    <div>
      <div
        onClick={[Function]}
        role="presentation"
      />
      <input
        onFocus={[Function]}
        readOnly={true}
        type="datetime"
        value=""
      />
    </div>
    <button
      onClick={[Function]}
    >
      <Component
        icon="times"
      />
    </button>
  </div>
  <Component
    isOpened={false}
  >
    <button
      onClick={[Function]}
    />
    <Component
      dateFormat={true}
      onChange={[Function]}
      open={true}
      timeFormat={true}
    />
    <Component
      onClick={[Function]}
      style="brand"
    />
  </Component>
</div>
`;
